crates/types/src/vid.rs

```diff
@@ -0,0 +1,254 @@
+//! This module provides:
+//! - an opaque constructor [`vid_scheme`] that returns a new instance of a
+//!   VID scheme.
+//! - type aliases [`VidCommitment`], [`VidCommon`], [`VidShare`]
+//!   for [`VidScheme`] assoc types.
+//!
+//! Purpose: the specific choice of VID scheme is an implementation detail.
+//! This crate and all downstream crates should talk to the VID scheme only
+//! via the traits exposed here.
+
+use ark_bls12_381::Bls12_381;
+use jf_primitives::{
+    pcs::{checked_fft_size, prelude::UnivariateKzgPCS, PolynomialCommitmentScheme},
+    vid::{
+        advz::{
+            payload_prover::{LargeRangeProof, SmallRangeProof},
+            Advz,
+        },
+        payload_prover::{PayloadProver, Statement},
+        precomputable::Precomputable,
+        VidDisperse, VidResult, VidScheme,
+    },
+};
+use serde::{Deserialize, Serialize};
+use sha2::Sha256;
+use std::{fmt::Debug, ops::Range};
+
+/// VID scheme constructor.
+///
+/// Returns an opaque type that impls jellyfish traits:
+/// [`VidScheme`], [`PayloadProver`], [`Precomputable`].
+///
+/// # Rust forbids naming impl Trait in return types
+///
+/// Due to Rust limitations the return type of [`vid_scheme`] is a newtype
+/// wrapper [`VidSchemeType`] that impls the above traits.
+///
+/// We prefer that the return type of [`vid_scheme`] be `impl Trait` for the
+/// above traits. But the ability to name an impl Trait return type is
+/// currently missing from Rust:
+/// - [Naming impl trait in return types - Impl trait initiative](https://rust-lang.github.io/impl-trait-initiative/explainer/rpit_names.html)
+/// - [RFC: Type alias impl trait (TAIT)](https://github.com/rust-lang/rfcs/blob/master/text/2515-type_alias_impl_trait.md)
+///
+/// # Panics
+/// When the construction fails for the underlying VID scheme.
+#[must_use]
+pub fn vid_scheme(num_storage_nodes: usize) -> VidSchemeType {
+    // TODO use a proper SRS
+    // https://github.com/EspressoSystems/HotShot/issues/1686
+    let srs = {
+        let mut rng = jf_utils::test_rng();
+        UnivariateKzgPCS::<E>::gen_srs_for_testing(
+            &mut rng,
```

Comment:
`test_rng` is deterministic, so this should produce the same SRS each time. We've been doing it like this since the dawn of time so if it were a problem then we'd know about it by now. That said, we definitely need to stop regenerating the SRS for each VID scheme. 😬 This is something on the crypto team's plate. Related: https://github.com/EspressoSystems/HotShot/issues/1686